File: whoami.go
package main

import "os"

var authTopic = &Topic{
	Name:        "auth",
	Description: "authentication (login/logout)",
}

var whoamiTopic = &Topic{
	Name:   "whoami",
	Hidden: true,
}

var whoamiCmd = &Command{
	Topic:       "whoami",
	Description: "display your Heroku login",
	Help: `Example:

  $ heroku auth:whoami
	email@example.com

	whoami will return nonzero status if not logged in:

  $ heroku auth:whoami
	not logged in
	$ echo $?
	100`,
	Run: whoami,
}

var whoamiAuthCmd = &Command{
	Topic:       "auth",
	Command:     "whoami",
	Description: "display your Heroku login",
	Default:     true,
	Help: `Example:

  $ heroku auth:whoami
	email@example.com

	whoami will return nonzero status if not logged in:

  $ heroku auth:whoami
	not logged in
	$ echo $?
	100`,
	Run: whoami,
}

func whoami(ctx *Context) {
	if os.Getenv("HEROKU_API_KEY") != "" {
		Warn("HEROKU_API_KEY is set")
	}

	// don't use needsToken since this should fail if
	// not logged in. Should not show a login prompt.
	ctx.APIToken = apiToken()

	if ctx.APIToken == "" {
		Println("not logged in")
		Exit(100)
	}

	user := getUserFromToken(ctx.APIToken)
	if user == "" {
		Println("not logged in")
		Exit(100)
	}
	Println(user)
}

File: client/lib/commands/auth/whoami.js
'use strict'

const co = require('co')
const cli = require('heroku-cli-util')

function * run () {
  const Heroku = require('heroku-client')

  if (process.env.HEROKU_API_KEY) {
    cli.warn('HEROKU_API_KEY is set. Not using netrc credentials.')
  }
  let token = cli.auth.token()
  if (!token) cli.exit(1, 'not logged in')
  let heroku = new Heroku({token})
  let account = yield heroku.get('/account')
  cli.log(account.email)
}

module.exports = {
  topic: 'auth',
  command: 'whoami',
  description: 'display the logged in Heroku user',
  run: co.wrap(run)
}
